spec/index.bs

```diff
@@ -4440,13 +4451,19 @@ dictionary GPUShaderModuleDescriptor : GPUObjectDescriptorBase {
 {{GPUShaderModuleDescriptor/sourceMap}}, if defined, MAY be interpreted as a
 source-map-v3 format.
 Source maps are optional, but serve as a standardized way to support dev-tool
-integration such as source-language debugging. [[SourceMap]]
+integration such as source-language debugging [[SourceMap]].
+WGSL names in source maps are interpreted under [=UAX15 Normalization Forms|Normalization Form C (NFC)=]
```

Comment:
I don't know what it means to interpret something under a normalization form.

Reply:
@litherum Should it just say "normalized with" or "parsed as"?

Reply:
Or maybe "used by their normalized form per [=UAX15 Normalization Forms|Normalization Form C (NFC)=]"

Reply:
The WGSL spec already has a section that says "To parse a WGSL program: 1. Remove comments ... 2. ... group the remaining characters into tokens and blankspace ... etc."

I think you should just add a step in that list that is "normalize the string to NFC form." Also it would be useful to add a note saying A) why we are doing this, and B) that this is observable in the line/column fields of error messages. Also it might be useful to add an example of how a JS page can normalize their code themselves to match the error index fields to their string.

And similar for JavaScript functions in the API - the descriptions of what the functions do can just say "step 1: normalize the entry point name string to NFC form" or something like that.

Reply:
@litherum Did the consensus reach to having normalized impact developer output in error messages? I think the agreement was normalization only taking place in comparison between strings (though not actually comparison but rather at the stage of hashing most probably).

Inclusion in the steps definitely makes sense; I'll apply the change.

Reply:
(which is why the following line is `normalization does not impact line/column data ... to avoid misalignments with the original`)

Reply:
I don't think we had a consensus, but my opinion is that it's better to have normalization occur during identifier resolution, rather than prior to parsing, so that it doesn't affect error message locations.

I think if it affects error message locations it's harmful to internationalization:
> this would be rather unfortunate because it means that users of the compilationInfo() API would NOT be internationalized by default [as they have to explicitly normalize the string].
https://github.com/gpuweb/gpuweb/issues/1160#issuecomment-1034388178

Reply:
Ah ok I thought we had consensus

Reply:
I agree with @kainino0x that the normalization should not impact error messages

Reply:
There is a potential perf benefit, by the way, of deferring normalization until comparison time: from https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/unorm2_8h.html#aee41897bd6c4ff7dc7a8db6d37efdcbd

> This function compares strings incrementally instead of normalizing (and optionally case-folding) both strings entirely, improving performance significantly.

I thought the reason for the idea of normalizing early was that Dawn didn’t want to link with ICU?

Reply:
> I thought the reason for the idea of normalizing early was that Dawn didn’t want to link with ICU?

Yes, but that was before I realized there was a concrete problem with it. In Chromium, linking with ICU is not a concern, so we'll just have to figure out how to get the linking set up, and perhaps consider having the behavior differ slightly when using Dawn outside of Chromium.

Reply:
I think you should use the verb “to normalize”

Reply:
We may need a new section that describes how string comparisons in WGSL work, and have other places in the spec link to that section.

Reply:
@litherum I did start one tagged #identifier-normalization, maybe extending that?

Reply:
@litherum spec now says `normalized to their NFC form`

Reply:
🆒